src/main/java/org/cirdles/topsoil/app/CloseTabEventHandler.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright 2015 CIRDLES.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.cirdles.topsoil.app;
+
+import java.util.Optional;
+import java.util.function.Consumer;
+import javafx.event.Event;
+import javafx.event.EventHandler;
+import javafx.scene.control.Alert;
+import javafx.scene.control.ButtonType;
+import javafx.scene.control.Tab;
+import org.cirdles.topsoil.dataset.DatasetResource;
+import org.cirdles.topsoil.dataset.DatasetResourceManager;
+
+/**
+ *
+ * @author parizotclement
+ */
+public class CloseTabEventHandler implements EventHandler<Event> {
+
+    private final Tab tab;
+    private final DatasetResourceManager datasetResourceManager;
+    private final Consumer<Boolean> saveDataTable;
+
+    public CloseTabEventHandler(Tab tab, DatasetResourceManager datasetResourceManager, Consumer<Boolean> callback) {
+        this.tab = tab;
+        this.datasetResourceManager = datasetResourceManager;
+        this.saveDataTable = callback;
+    }
+
+    @Override
+    public void handle(Event event) {
+        if (tab.getContent() instanceof TSVTable) {
+
+            tab.getTabPane().getSelectionModel().select(tab);
+
+            TSVTable table = (TSVTable) tab.getContent();
+            Optional<DatasetResource> datasetResource = datasetResourceManager.getDatasetResource(table.getDataset());
+
+            if (datasetResource.isPresent()) {
+                datasetResource.ifPresent(
+                        resource -> resource.close()
+                );
+            } else {
+                confirmationDialog().ifPresent((ButtonType choice) -> {
+                    switch (choice.getText()) {
+                        case "Save":
+                            saveDataTable.accept(true);
+                            break;
+                        case "Cancel":
+                            event.consume();
+                            break;
+                    }
+                });
```

Comment:
It's hard to follow because you have a `case` inside of an `if`/`else` inside of an `if`. There are like 12 paths to think through here.